### PR TITLE
CRC64 perf improvements from Redis patches

### PR DESCRIPTION
**Pull request by josiahcarlson**
Improve the performance of crc64 for large batches by processing large number of bytes in parallel and combining the results. 

## Performance 
* 53-73% faster on Xeon 2670 v0 @ 2.6ghz
* 2-2.5x faster on Core i3 8130U @ 2.2 ghz
* 1.6-2.46 bytes/cycle on i3 8130U
* likely >2x faster than crcspeed on newer CPUs with more resources than a 2012-era Xeon 2670
* crc64 combine function runs in <50 nanoseconds typical with vector + cache optimizations (~8 *microseconds* without vector optimizations, ~80 *microseconds without cache, the combination is extra effective)
* still single-threaded
* valkey-server test crc64 --help (requires `make distclean && make SERVER_TEST=yes`)

### CODE REVIEW

**Comment by zuiderkwast** [comment]
What do we use crc64 for? RDB files?

I want to understand the practical benefit.

**Comment by madolson** [comment]
> What do we use crc64 for? RDB files?

RDB and slot migration. (Since the dump + restore includes the CRC checksumming). I mentioned before improving crc16 would be a bigger win for cluster mode.

**Comment by zuiderkwast** [comment]
OK. It would be impossible to change to SHA1 or something else for RDB, DUMP/RESTORE, right? Because of backwards compatibility.

I think we should merge this then. It's more complex code, but as long as it's well tested and isolated in its own files, that's no problem to me.

File: src/crc64.c [hunk 1]
```diff
@@ -67,14 +67,32 @@ static uint64_t crc64_table[8][256] = {{0}};
  * \return             The reflected data.
  *****************************************************************************/
 static inline uint_fast64_t crc_reflect(uint_fast64_t data, size_t data_len) {
-    uint_fast64_t ret = data & 0x01;
+    /* only ever called for data_len == 64 in this codebase
```
Comment (zuiderkwast) [inline src/crc64.c:71]: Does the code assume data_len == 64?

* If yes, we should add an assert for that
* If no, why mention it at all?

File: README.md [hunk 2]
```diff
@@ -54,6 +54,10 @@ installed):
     % ./utils/gen-test-certs.sh
     % ./runtest --tls
 
+To build and test crc 64 performance improvements:
+
+    % make distclean && make SEVER_TEST=yes
```
Comment (zuiderkwast) [inline README.md]: Typo?

```suggestion
    % make distclean && make SERVER_TEST=yes
```

Btw, if we document `src/valkey-server test` for crc64, we should probably document it for all of the tests that run in this way.

Maybe it's enough to document it a comment in the Makefile? The README only documents what can be useful for end users. It doesn't include running with sanitizers and a lot of other things.

File: src/Makefile [hunk 3]
```diff
@@ -36,6 +36,9 @@ NODEPS:=clean distclean
 
 # Default settings
 STD=-pedantic -DSERVER_STATIC=''
+ifeq ($(SERVER_TEST),yes)
+	STD +=-DSERVER_TEST=1
+endif
```
Comment (zuiderkwast) [inline src/Makefile]: It's good to have this make variable, but I think it's not right to add it to `STD`. We can add it to `FINAL_CFLAGS` instead, around 100 lines down .

File: src/crccombine.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	 in a product, an acknowledgment in the product documentation would be
+	 appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+	 misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+
+  Mark Adler
+  madler@alumni.caltech.edu
+*/
+
```
Comment (zuiderkwast) [inline src/crccombine.c:40]: This is a new file that we link against. Is it copied from Mark Adler's code? We didn't we need that before?

I think it'd be useful to have a very short comment what this file does, in relation to the other crc files.

**Review by zuiderkwast — COMMENTED** [review]
Superficial review. I didn't proof read the bit fiddling, but I'm not intending to.

When producing or loading a large RDB dump, how much of the CPU time is CRC64 computation and how much is other stuff?

If the CRC computation is haft of the CPU time, it's good, but if the CRC computation is only 1%, then this optimization is useless IMO.

**Comment by zuiderkwast** [comment]
I found your blog post. :smiley: https://www.dr-josiah.com/2023/04/crc64-in-redis-is-getting-even-faster.html

File: README.md (hunk 2 above)
Comment (josiahcarlson) [reply README.md]: I will update the docs.

**Comment by josiahcarlson** [comment]
> When producing or loading a large RDB dump, how much of the CPU time is CRC64 computation and how much is other stuff?

Let me rephrase your question; why is doing CRC64 important for Redis?

Before loading a snapshot from disk, or received over the wire from a master server, the loading server will take a pass over the whole RDB file, and call out roughly once every 2 megabytes by default (see `loading_process_events_interval_bytes`). These changes save ~.5 seconds per gigabyte of snapshot to be loaded.

When creating RDB snapshots, any data segment written gets a call to CRC64 for up to 2 megabytes by default (also uses `loading_process_events_interval_bytes`), but typically closer to whatever the data storage size. So maybe tens of bytes at present.

That said, with additional small changes, you can update on write the same 2 meg segment all the time (instead of at maximum), and the effects of this (along with eliminating fork overhead) can be seen in 20 seconds of reduced startup / snapshot time for a 10 gig resident dataset from the last slide here: https://www.slideshare.net/secret/sLTnWqcPV7G0HK , of which ~8-10 seconds can be attributed to CRC64 performance improvements.

**Comment by zuiderkwast** [comment]
Sounds good. I'm convinced. :)

FYI: You need Signed-off-by also for commits done using suggestions in the web gui. You can add that using `git rebase --signoff HEAD~1` and then force-push. (Replace 1 with N to add signoff to the last N commits.)

File: src/crc64.c (hunk 1 above)
Comment (josiahcarlson) [reply src/crc64.c:71]: * Does not assume the data_len is 64.
* Does not require data_len be 64, though has a practical limitation that `0 <= data_len <= 64` - I will add those asserts.
* This is a usage note to inform future pager rotations (sometimes me) woken up at 3AM on a Sunday.

Comment (josiahcarlson) [reply src/crc64.c:71]: With a size_t, only the `<= 64` is necessary for proper functioning here, I've added the assert.

File: README.md (hunk 2 above)
Comment (madolson) [reply README.md]: For reference, I'm trying to move away from this testing strategy to one that can be called without rebuilding: https://github.com/valkey-io/valkey/pull/344. I would actually prefer just leaving them undocumented for now.

File: src/crc64.c [hunk 5]
```diff
@@ -28,6 +28,7 @@
 
 #include "crc64.h"
 #include "crcspeed.h"
+#include "serverassert.h"
```
Comment (zuiderkwast) [inline src/crc64.c:31]: It's hurts a bit to see this, since the CRC files were independent from the rest of the code base until now.

I see all other copied libraries have this though, so I guess it's the way it's done in this code base.

Was that your conclusion?

Comment (madolson) [reply src/crc64.c:31]: Yeah. It's the way we override the normal "assert" with the valkey server assert, so we get our nice pretty stack traces :). Otherwise it just exits (or until very recently, it was just ignoring it).

**Review by madolson — APPROVED** [review]

**Comment by madolson** [comment]
Assuming tests pass, I'm happy with it.

**Review by zuiderkwast — APPROVED** [review]
+1

File: src/crc64.c (hunk 5 above)
Comment (zuiderkwast) [reply src/crc64.c:31]: If anyone wants to use it outside of valkey, it's easy enough to replace this line with a regular `<assert.h>`.

File: src/crccombine.c [hunk 6]
```diff
@@ -0,0 +1,172 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <strings.h>
+#include <immintrin.h>
```
Comment (madolson) [inline src/crccombine.c:5]: I suppose I didn't consider this originally, but I guess this isn't supported on ARM. I assume there is likely a version that implements. We could also just make this an x86 only optimization for now.

Comment (josiahcarlson) [reply src/crccombine.c:5]: Added some macros and a non-vector (but faster than bitwise) multiply function from the original patchset. Assuming everyone has \_ARM\_ defined as 0 or 1, this should now be cross-platform. (it exists on Linux x86 as 0)

I also bumped the cutoffs based on my experience with the slower combine functions.

Comment (zuiderkwast) [reply src/crccombine.c:5]: Valkey can be compiled for some weird systems so I'm not so sure we can assume `_ARM_` is always defined. (We have some ifdefs for Solaris and multiple \*BSDs.)

Better write `#if !defined(_ARM_) || !_ARM_` just in case, to avoid warnings when compiling with `-Wundef`.

Comment (madolson) [reply src/crccombine.c:5]: We defined `__arm__` in config.h, maybe we import that and use that define.

Comment (zuiderkwast) [reply src/crccombine.c:5]: @madolson I like the separation so i'd prefer to avoid including config.h if we can.

Comment (madolson) [reply src/crccombine.c:5]: Is it all that important? They are our files that we are modifying in the end.

File: src/crccombine.c [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+		cache_num = (cache_num + 1) & 63;
+		/* if no more bits set, then done */
+	} while (len2 != 0);
+
+	/* return combined crc */
+	crc1 ^= crc2;
+	return crc1;
+}
+
+#undef CRC_MULTIPLY
```
Comment (madolson) [inline src/crccombine.c]:
```suggestion
#undef CRC_MULTIPLY

```

File: src/crcspeed.c [hunk 8]
```diff
@@ -104,45 +122,152 @@ void crcspeed16big_init(crcfn16 fn, uint16_t big_table[8][256]) {
...
+static size_t CRC64_DUAL_CUTOFF = (1024);
+#endif
+
+
+void set_crc64_cutoffs(size_t dual_cutoff, size_t tri_cutoff) {
+    CRC64_DUAL_CUTOFF = dual_cutoff;
+    CRC64_TRI_CUTOFF = tri_cutoff;
+}
+
+
```
Comment (madolson) [inline src/crcspeed.c]:
```suggestion
```
Just making a comment so I can commit it to get github to pull a fix I made to unstable.

File: src/crccombine.c (hunk 6 above)
Comment (zuiderkwast) [reply src/crccombine.c:5]: It's not that important but it seemed easy to avoid.

Comment (zuiderkwast) [reply src/crccombine.c:5]: It would hit us if we (or someone else) ever want to use this code for something else, like a separate rdb tool. That they're our files is not a good reason to add unnecessary coupling imo.

Comment (madolson) [reply src/crccombine.c:5]: I just realized that "serverassert.h" imports config.h, so if we want to be careful about writing portable code, we need to make sure to avoid the config.h defines since it will seemingly work until someone tries to do this replacement.

Comment (zuiderkwast) [reply src/crccombine.c:5]: Yes. :/ IMO it shouldn't. It's just for `likely` and `valkey_unreachable`, which only depend on the macro `__GNUC__`.

Comment (madolson) [reply src/crccombine.c:5]: Ok, going to close this since you proposed a fix here: https://github.com/valkey-io/valkey/pull/404.

**Comment by madolson** [comment]
Btw, I just want to test it on another arm machine. I'll do that tomorrow and merge it if the performance makes sense.

File: src/crcspeed.c [hunk 9]
```diff
@@ -104,45 +122,151 @@ void crcspeed16big_init(crcfn16 fn, uint16_t big_table[8][256]) {
...
+/* Variables so we can change for benchmarking; these seem to be fairly
+ * reasonable for Intel CPUs made since 2010. Please adjust as necessary if
+ * or when your CPU has more load / execute units. We've written benchmark code
+ * to help you tune your platform, see crc64Test. */
+#if defined(__i386__) || defined(__X86_64__)
+static size_t CRC64_TRI_CUTOFF = (2*1024);
+static size_t CRC64_DUAL_CUTOFF = (128);
+#else
+static size_t CRC64_TRI_CUTOFF = (16*1024);
+static size_t CRC64_DUAL_CUTOFF = (1024);
```
Comment (madolson) [inline src/crcspeed.c:165]: Testing on a graviton3, these seem about right.

File: src/crccombine.c [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+	/*
+	 * Uses xmm registers on x86, works basically everywhere fast, doing
+	 * cycles of movqda, mov, shr, pand, and, pxor, at least on gcc 8.
+	 * Is 9-11x faster than original.
+	 * E5-2670 ~29us to extend ~1 meg 64 bit hash
+	 * i3-8130U ~22us to extend ~1 meg 64 bit hash
+	 */
+	v2uq sum = {0, 0},
+		*mv2 = (v2uq*)mat;
+    /* this table allows us to eliminate conditions during gf2_matrix_times_vec2() */
```
Comment (madolson) [inline src/crccombine.c]:
```suggestion
	/* this table allows us to eliminate conditions during gf2_matrix_times_vec2() */
```

**Review by madolson — APPROVED** [review]
LGTM, builds and some quick tests show it's faster than crcspeed.

**Comment by yawaramin** [comment]
Please see https://github.com/redis/redis/pull/13638#issuecomment-2874530225

**Comment by madolson** [comment]
@yawaramin It's BSD code, they are allowed to do whatever they want with it as long as they include the attribution of the code.

**Comment by yawaramin** [comment]
The licenses are not the primary issue here. Redis has a CLA which means contributors have to give away copyright of the code to Redis. So they no longer have ownership of the code and can't certify it to Valkey as required by its DCO.

**Comment by madolson** [comment]
You can take a look at Redis' CLA here, https://redis.io/legal/redis-software-grant-and-contributor-license-agreement/. You are only granting Redis a license to prepare derivative works by signing the CLA, you're not handing over the copyright.
